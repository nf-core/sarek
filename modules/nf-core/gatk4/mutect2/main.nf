process GATK4_MUTECT2 {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/gatk4:4.5.0.0--py36hdfd78af_0':
        'biocontainers/gatk4:4.5.0.0--py36hdfd78af_0' }"

    input:
    tuple val(meta), path(input), path(input_index), path(intervals)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(fai)
    tuple val(meta4), path(dict)
    path(germline_resource)
    path(germline_resource_tbi)
    path(panel_of_normals)
    path(panel_of_normals_tbi)

    output:
    tuple val(meta), path("*.vcf.gz")     , emit: vcf
    tuple val(meta), path("*.tbi")        , emit: tbi
    tuple val(meta), path("*.stats")      , emit: stats
    tuple val(meta), path("*.f1r2.tar.gz"), optional:true, emit: f1r2
    path "versions.yml"                   , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def inputs = input.collect{ "--input $it"}.join(" ")
    def interval_command = intervals ? "--intervals $intervals" : ""
    def pon_command = panel_of_normals ? "--panel-of-normals $panel_of_normals" : ""
    def gr_command = germline_resource ? "--germline-resource $germline_resource" : ""

    def avail_mem = 3072
    if (!task.memory) {
        log.info '[GATK Mutect2] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
    } else {
        avail_mem = (task.memory.mega*0.8).intValue()
    }
    """
    gatk --java-options "-Xmx${avail_mem}M -XX:-UsePerfData" \\
        Mutect2 \\
        $inputs \\
        --output ${prefix}.vcf.gz \\
        --reference $fasta \\
        $pon_command \\
        $gr_command \\
        $interval_command \\
        --tmp-dir . \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.vcf.gz
    touch ${prefix}.vcf.gz.tbi
    touch ${prefix}.vcf.gz.stats
    touch ${prefix}.f1r2.tar.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
    END_VERSIONS
    """
}
