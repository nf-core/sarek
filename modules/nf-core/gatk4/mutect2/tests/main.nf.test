nextflow_process {

    name "Test Process GATK4_MUTECT2"
    script "../main.nf"
    process "GATK4_MUTECT2"

    tag "modules"
    tag "modules_nfcore"
    tag "gatk4"
    tag "gatk4/mutect2"

    test("tumor_normal_pair") {
        config "./pair.config"
        when {
            process {
            """
            input[0] = [
                [
                    id:'test',
                    normal_id:'normal',
                    tumor_id:'tumour'
                ],
                [
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.recalibrated.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.recalibrated.sorted.bam', checkIfExists: true)
                ],
                [
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.recalibrated.sorted.bam.bai', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.recalibrated.sorted.bam.bai', checkIfExists: true)
                ],
                []
            ]
            input[1] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta', checkIfExists: true)
            ]
            input[2] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta.fai', checkIfExists: true)
            ]
            input[3] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.dict', checkIfExists: true)
            ]
            input[4] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz', checkIfExists: true)
            input[5] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz.tbi', checkIfExists: true)
            input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz', checkIfExists: true)
            input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz.tbi', checkIfExists: true)
            """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.vcf.collect { file(it[1]).getName() },
                        process.out.tbi.collect { file(it[1]).getName() },
                        process.out.stats,
                        process.out.f1r2,
                        process.out.versions,
                    ).match()
                }
            )
        }
    }

    test("tumor_normal_pair_f1r2") {
        config "./f1r2.config"
        when {
            process {
            """
            input[0] = [
                [
                    id:'test',
                    normal_id:'normal',
                    tumor_id:'tumour'
                ],
                [
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.recalibrated.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.recalibrated.sorted.bam', checkIfExists: true)
                ],
                [
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.recalibrated.sorted.bam.bai', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.recalibrated.sorted.bam.bai', checkIfExists: true)
                ],
                []
            ]
            input[1] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta', checkIfExists: true)
            ]
            input[2] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta.fai', checkIfExists: true)
            ]
            input[3] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.dict', checkIfExists: true)
            ]
            input[4] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz', checkIfExists: true)
            input[5] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz.tbi', checkIfExists: true)
            input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz', checkIfExists: true)
            input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz.tbi', checkIfExists: true)
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.vcf.collect { file(it[1]).getName() },
                        process.out.tbi.collect { file(it[1]).getName() },
                        process.out.stats,
                        process.out.f1r2.collect { file(it[1]).getName() },
                        process.out.versions
                    ).match()
                }
            )
        }
    }
    test("tumor_single"){
        when {
            process {
            """
            input[0] = [
                [ id:'test'],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.recalibrated.sorted.bam', checkIfExists: true)],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.recalibrated.sorted.bam.bai', checkIfExists: true)],
                []
            ]
            input[1] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta', checkIfExists: true)
            ]
            input[2] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta.fai', checkIfExists: true)
            ]
            input[3] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.dict', checkIfExists: true)
            ]
            input[4] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz', checkIfExists: true)
            input[5] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz.tbi', checkIfExists: true)
            input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz', checkIfExists: true)
            input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz.tbi', checkIfExists: true)
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.vcf.collect { file(it[1]).getName() },
                        process.out.tbi.collect { file(it[1]).getName() },
                        process.out.stats,
                        process.out.f1r2,
                        process.out.versions
                    ).match()
                }
            )
        }
    }
    test("cram_input"){
        when {
            process{
            """
            input[0] = [
                [ id:'test'],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test2.paired_end.recalibrated.sorted.cram', checkIfExists: true)],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test2.paired_end.recalibrated.sorted.cram.crai', checkIfExists: true)],
                []
            ]
            input[1] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta', checkIfExists: true)
            ]
            input[2] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta.fai', checkIfExists: true)
            ]
            input[3] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.dict', checkIfExists: true)
            ]
            input[4] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz', checkIfExists: true)
            input[5] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz.tbi', checkIfExists: true)
            input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz', checkIfExists: true)
            input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz.tbi', checkIfExists: true)
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.vcf.collect { file(it[1]).getName() },
                        process.out.tbi.collect { file(it[1]).getName() },
                        process.out.stats,
                        process.out.f1r2,
                        process.out.versions
                    ).match()
                }
            )
        }
    }

    test("generate_pon") {
        when {
            process {
            """
            input[0] = [
                [ id:'test'],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.recalibrated.sorted.bam', checkIfExists: true)],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.recalibrated.sorted.bam.bai', checkIfExists: true)],
                []
            ]
            input[1] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta', checkIfExists: true)
            ]
            input[2] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta.fai', checkIfExists: true)
            ]
            input[3] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.dict', checkIfExists: true)
            ]
            input[4] = []
            input[5] = []
            input[6] = []
            input[7] = []
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.vcf.collect { file(it[1]).getName() },
                        process.out.tbi.collect { file(it[1]).getName() },
                        process.out.stats,
                        process.out.f1r2,
                        process.out.versions
                    ).match()
                }
            )
        }
    }

    test("mitochondria"){
        when {
            process {
            """
            input[0] = [
                [ id:'test'],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/mitochon_standin.recalibrated.sorted.bam', checkIfExists: true)],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/mitochon_standin.recalibrated.sorted.bam.bai', checkIfExists: true)],
                [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.bed', checkIfExists: true)]
            ]
            input[1] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
            ]
            input[2] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
            ]
            input[3] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
            ]
            input[4] = []
            input[5] = []
            input[6] = []
            input[7] = []
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.vcf.collect { file(it[1]).getName() },
                        process.out.tbi.collect { file(it[1]).getName() },
                        process.out.stats,
                        process.out.f1r2,
                        process.out.versions
                    ).match()
                }
            )
        }
    }

    test("tumor_normal_pair_f1r2_stubs"){
        options "-stub-run"
        when {
            process {
            """
            input[0] = [
                [
                    id:'test',
                    normal_id:'normal',
                    tumor_id:'tumour'
                ],
                [
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.recalibrated.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.recalibrated.sorted.bam', checkIfExists: true)
                ],
                [
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.recalibrated.sorted.bam.bai', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.recalibrated.sorted.bam.bai', checkIfExists: true)
                ],
                []
            ]
            input[1] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta', checkIfExists: true)
            ]
            input[2] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.fasta.fai', checkIfExists: true)
            ]
            input[3] = [
                [ id:'genome' ],
                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/sequence/genome.dict', checkIfExists: true)
            ]
            input[4] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz', checkIfExists: true)
            input[5] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/gnomAD.r2.1.1.vcf.gz.tbi', checkIfExists: true)
            input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz', checkIfExists: true)
            input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/chr21/germlineresources/mills_and_1000G.indels.hg38.vcf.gz.tbi', checkIfExists: true)
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.vcf.collect { file(it[1]).getName() },
                        process.out.tbi.collect { file(it[1]).getName() },
                        process.out.stats.collect { file(it[1]).getName() },
                        process.out.f1r2.collect { file(it[1]).getName() },
                        process.out.versions.collect { file(it[1]).getName() }
                    ).match()
                }
            )
        }

    }

}
