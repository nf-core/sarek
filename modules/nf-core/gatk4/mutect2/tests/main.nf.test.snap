{
    "tumor_normal_pair_f1r2_stubs": {
        "content": [
            [
                "test.vcf.gz"
            ],
            [
                "test.vcf.gz.tbi"
            ],
            [
                "test.vcf.gz.stats"
            ],
            [
                "test.f1r2.tar.gz"
            ],
            [
                "h"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.2"
        },
        "timestamp": "2024-03-21T10:14:45.599103891"
    },
    "generate_pon": {
        "content": [
            [
                "test.vcf.gz"
            ],
            [
                "test.vcf.gz.tbi"
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.stats:md5,b569ce66bbffe9588b3d221e821023ee"
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,d94731c50c20569fe9896235a843f382"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.2"
        },
        "timestamp": "2024-03-20T15:57:18.264453766"
    },
    "mitochondria": {
        "content": [
            [
                "test.vcf.gz"
            ],
            [
                "test.vcf.gz.tbi"
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.stats:md5,4f77301a125913170b8e9e7828b4ca3f"
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,d94731c50c20569fe9896235a843f382"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.2"
        },
        "timestamp": "2024-03-20T16:05:47.668766905"
    },
    "cram_input": {
        "content": [
            [
                "test.vcf.gz"
            ],
            [
                "test.vcf.gz.tbi"
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.stats:md5,55ed641e16089afb33cdbc478e202d3d"
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,d94731c50c20569fe9896235a843f382"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.2"
        },
        "timestamp": "2024-03-20T15:52:27.894730554"
    },
    "tumor_single": {
        "content": [
            [
                "test.vcf.gz"
            ],
            [
                "test.vcf.gz.tbi"
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.vcf.gz.stats:md5,55ed641e16089afb33cdbc478e202d3d"
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,d94731c50c20569fe9896235a843f382"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.2"
        },
        "timestamp": "2024-03-20T15:43:28.935723443"
    },
    "tumor_normal_pair": {
        "content": [
            [
                "test.vcf.gz"
            ],
            [
                "test.vcf.gz.tbi"
            ],
            [
                [
                    {
                        "id": "test",
                        "normal_id": "normal",
                        "tumor_id": "tumour"
                    },
                    "test.vcf.gz.stats:md5,17d2091015d04cbd4a26b7a67dc659e6"
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,d94731c50c20569fe9896235a843f382"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.2"
        },
        "timestamp": "2024-03-20T15:31:31.913366311"
    },
    "tumor_normal_pair_f1r2": {
        "content": [
            [
                "test.vcf.gz"
            ],
            [
                "test.vcf.gz.tbi"
            ],
            [
                [
                    {
                        "id": "test",
                        "normal_id": "normal",
                        "tumor_id": "tumour"
                    },
                    "test.vcf.gz.stats:md5,17d2091015d04cbd4a26b7a67dc659e6"
                ]
            ],
            [
                "test.f1r2.tar.gz"
            ],
            [
                "versions.yml:md5,d94731c50c20569fe9896235a843f382"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.2"
        },
        "timestamp": "2024-03-21T09:45:52.321385704"
    }
}